File: Medium/840. Magic Squares In Grid.java
class Solution {
    public int numMagicSquaresInside(int[][] grid) {
        final int r = grid.length, c = grid[0].length;
        int res = 0;
        for (int i = 0; i < r - 2; i++) {
            for (int j = 0; j < c - 2; j++) {
                int sum = grid[i][j] + grid[i][j + 1] + grid[i][j + 2];
                if (isMagic(grid, i, j, sum)) {
                    res++;
                }
            }
        }
        return res;
    }

    private boolean isMagic(int[][] grid, int r, int c, int sum) {
        final var cnt = new int[10];
        for (int i = r + 2; i >= r; i--) {
            for (int j = c + 2; j >= c; j--) {
                if (grid[i][j] > 9 || grid[i][j] < 1) {
                    return false;
                }
                if (++cnt[grid[i][j]] > 1) {
                    return false;
                }
            }
        }
        for (int i = r + 1; i <= r + 2; i++) {
            if (sum != grid[i][c] + grid[i][c + 1] + grid[i][c + 2]) {
                return false;
            }
        }
        for (int j = c; j <= c + 2; j++) {
            if (sum != grid[r][j] + grid[r + 1][j] + grid[r + 2][j]) {
                return false;
            }
        }
        if (sum != grid[r][c] + grid[r + 1][c + 1] + grid[r + 2][c + 2]) {
            return false;
        }
        if (sum != grid[r][c + 2] + grid[r + 1][c + 1] + grid[r + 2][c]) {
            return false;
        }
        return true;
    }
}

File: Medium/840. Magic Squares In Grid.py
from typing import List


class Solution:
    def numMagicSquaresInside(self, grid: List[List[int]]) -> int:
        R, C = len(grid), len(grid[0])
        magic_cnt = 0

        def isMagic(r: int, c: int) -> bool:
            if grid[r + 1][c + 1] != 5:
                return False
            dig_sum1 = grid[r][c] + grid[r + 1][c + 1] + grid[r + 2][c + 2]
            dig_sum2 = grid[r][c + 2] + grid[r + 1][c + 1] + grid[r + 2][c]
            if  dig_sum1 != dig_sum2:
                return False
            num_set = [False] * 10
            row_sum = [0] * 3
            col_sum = [0] * 3
            for u in range(3):
                for v in range(3):
                    num = grid[r + u][c + v]
                    if not (0 < num < 10) or num_set[num]:
                        return False
                    num_set[num] = True
                    row_sum[u] += num
                    col_sum[v] += num
            return row_sum[0] == row_sum[1] == row_sum[2] == col_sum[0] == col_sum[1] == col_sum[2] == dig_sum1

        for r in range(R - 2):
            for c in range(C - 2):
                if isMagic(r, c):
                    magic_cnt += 1
                    
        return magic_cnt